### PR TITLE
feat: add b01 Q7 basic getter support

### PR DESCRIPTION
**Pull request by Lash-L**

### CODE REVIEW

File: roborock/devices/b01_channel.py [hunk 1]
```diff
@@ -2,26 +2,76 @@
...
+
+    def find_response(response_message: RoborockMessage) -> None:
+        """Handle incoming messages and resolve the future."""
+        try:
+            decoded_dps = decode_rpc_response(response_message)
+        except RoborockException as ex:
+            _LOGGER.info("Failed to decode b01 message: %s: %s", response_message, ex)
+            return
+
+        for _, dps_value in decoded_dps.items():
```
Comment (allenporter) [inline roborock/devices/b01_channel.py]:
```suggestion
        for dps_value in decoded_dps.values():
```

File: roborock/devices/b01_channel.py [hunk 2]
```diff
@@ -2,26 +2,76 @@
...
+                    continue
+
+                if isinstance(inner, dict) and inner.get("msgId") == msg_id:
+                    _LOGGER.debug("Received query response: %s", inner)
+                    data = inner.get("data")
+                    if isinstance(data, dict):
+                        result.update(data)
+                    finished.set()
+            else:
+                _LOGGER.debug("Received unexpected response: %s", dps_value)
```
Comment (allenporter) [inline roborock/devices/b01_channel.py]: Avoid the extra indent with early exit

```suggestion
            if not isinstance(dps_value, str):
                _LOGGER.debug("Received unexpected response: %s", dps_value)
                continue

            try:
                inner = json.loads(dps_value)
            except (json.JSONDecodeError, TypeError):
                _LOGGER.debug("Received unexpected response: %s", dps_value)
                continue

            if isinstance(inner, dict) and inner.get("msgId") == msg_id:
                _LOGGER.debug("Received query response: %s", inner)
                data = inner.get("data")
                if isinstance(data, dict):
                    result.update(data)
                finished.set()
```

File: roborock/devices/b01_channel.py [hunk 3]
```diff
@@ -2,26 +2,76 @@
...
     params: ParamsType,
-) -> None:
+) -> dict[str, Any]:
     """Send a command on the MQTT channel and get a decoded response."""
     _LOGGER.debug("Sending MQTT command: %s", params)
-    roborock_message = encode_mqtt_payload(dps, command, params)
-    await mqtt_channel.publish(roborock_message)
+    msg_id = str(get_next_int(100000000000, 999999999999))
+    roborock_message = encode_mqtt_payload(dps, command, params, msg_id)
+    finished = asyncio.Event()
```
Comment (allenporter) [inline roborock/devices/b01_channel.py]: `asyncio.Future` seems more appropriate here than  event + separate result object.

File: roborock/protocols/b01_protocol.py [hunk 4]
```diff
@@ -22,13 +22,13 @@
...
 
-def encode_mqtt_payload(dps: int, command: CommandType, params: ParamsType) -> RoborockMessage:
+def encode_mqtt_payload(dps: int, command: CommandType, params: ParamsType, msg_id: str) -> RoborockMessage:
     """Encode payload for B01 commands over MQTT."""
     dps_data = {
         "dps": {
             dps: {
                 "method": str(command),
-                "msgId": str(get_next_int(100000000000, 999999999999)),
+                "msgId": msg_id or str(get_next_int(100000000000, 999999999999)),
```
Comment (allenporter) [inline roborock/protocols/b01_protocol.py]: Why have this empty msg_id behavior?

File: tests/devices/traits/b01/test_init.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
+import json
```
Comment (allenporter) [inline tests/devices/traits/b01/test_init.py:1]: test_init.py?

Comment (Lash-L) [reply tests/devices/traits/b01/test_init.py:1]: Pytest was getting mad when it had a matching name to a01 test init for some reason

File: roborock/devices/b01_channel.py [hunk 6]
```diff
@@ -2,26 +2,76 @@
...
+                    inner = json.loads(dps_value)
+                except (json.JSONDecodeError, TypeError):
+                    _LOGGER.debug("Received unexpected response: %s", dps_value)
+                    continue
+
+                if isinstance(inner, dict) and inner.get("msgId") == msg_id:
+                    _LOGGER.debug("Received query response: %s", inner)
+                    data = inner.get("data")
+                    if isinstance(data, dict):
+                        result.update(data)
```
[the remark this answers is not in the document]
Comment (allenporter) [reply roborock/devices/b01_channel.py]: I think this is a valid issue. if we match the msg_id then we have the response and we should reply -- even with an exception -- rather than timeout.

File: tests/devices/traits/b01/test_init.py (hunk 5 above)
Comment (Lash-L) [reply tests/devices/traits/b01/test_init.py:1]: see here: https://github.com/Python-roborock/python-roborock/actions/runs/20200816620/job/57991726603?pr=662#step:5:41

Comment (allenporter) [reply tests/devices/traits/b01/test_init.py:1]: Ah i think i have seen this before and it can be resolved by adding an `__init__.py` file in the pytest directory.

Alternatively, we can stop using `__init_.py` in the traits since it may be hard to find the actual content in these subdirectories.

Comment (Lash-L) [reply tests/devices/traits/b01/test_init.py:1]: for now i have added a init as the alternative is a larger change

File: roborock/devices/b01_channel.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Lash-L) [reply roborock/devices/b01_channel.py]: raise an exception to clarify what is going wrong

**Review by allenporter — APPROVED** [review]

File: roborock/devices/b01_channel.py [hunk 7]
```diff
@@ -2,26 +2,76 @@
...
+
+    def find_response(response_message: RoborockMessage) -> None:
+        """Handle incoming messages and resolve the future."""
+        try:
+            decoded_dps = decode_rpc_response(response_message)
+        except RoborockException as ex:
+            _LOGGER.info("Failed to decode b01 message: %s: %s", response_message, ex)
+            return
+
+        for dps_value in decoded_dps.values():
```
Comment (allenporter) [inline roborock/devices/b01_channel.py:46]: For a future PR consider if `decode_rpc_response` should return a dict of parsed values similar to [a01 decode](https://github.com/Python-roborock/python-roborock/blob/4b4e5aa48fcc2ca2d091e3a2c3bc737b121cb878/roborock/protocols/a01_protocol.py#L54) then unit tests for corner cases can be more narrow without the full test setup.
